File: 0001-0100/037-sudoku-solver/037-sudoku-solver.cc
class Solution {
    bool backtracking(vector<vector<char>>& board) {
        const int n = 9;
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (board[i][j] != '.') continue;
                for (char c = '1'; c <= '9'; ++c) {
                    if (isValid(board, i, j, c)) {
                        board[i][j] = c;
                        if (backtracking(board)) return true;
                        board[i][j] = '.';
                    }
                }
                return false;
            }
        }
        return true;
    }

    bool isValid(vector<vector<char>> &board, int row, int col, char c) {
        for (int i = 0; i < 9; ++i) {
            if (board[i][col] == c) return false;
            if (board[row][i] == c) return false;
        }

        int startRow = row/3 * 3;
        int startCol = col/3 * 3;
        for (int i = startRow; i < startRow + 3; ++i) {
            for (int j = startCol; j < startCol + 3; ++j) {
                if (board[i][j] == c) return false;
            }
        }

        return true;
    }

public:
    void solveSudoku(vector<vector<char>>& board) {
        backtracking(board);
    }
};

File: 0001-0100/037-sudoku-solver/037-sudoku-solver.js
// 37. Sudoku Solver
// https://leetcode.com/problems/sudoku-solver/description/

/**
 * @param {character[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
var solveSudoku = function(board) {
  function putNumber(x, y) {
    x += Math.floor(y / n);
    y %= n;
    if (x > n - 1) return true;
    if (board[x][y] !== '.') {
      return putNumber(x, y + 1);
    }

    for (let i = 0; i < nums.length; i++) {
      const c = nums[i];
      const sid = Math.floor(x / 3) * 3 + Math.floor(y / 3);
      if (!row[x][c] && !col[y][c] && !square[sid][c]) {
        board[x][y] = c;
        row[x][c] = col[y][c] = square[sid][c] = true;
        if (putNumber(x, y + 1)) {
          return true;
        } else {
          row[x][c] = col[y][c] = square[sid][c] = false;
          board[x][y] = '.';
        }
      }
    }
    return false;
  }

  const nums = '123456789';
  const n = board.length;
  if (!n) return;

  const row = [], col = [], square = [];
  for (let i = 0; i < n; i++) {
    row.push(new Array(n + 1).fill(false));
    col.push(new Array(n + 1).fill(false));
    square.push(new Array(n + 1).fill(false));
  }

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      const c = board[i][j];
      if (c !== '.') {
        row[i][c] = true;
        col[j][c] = true;
        square[Math.floor(i / 3) * 3 + Math.floor(j / 3)][c] = true;
      }
    }
  }
  putNumber(0, 0);
};

const board = [
  ['5', '3', '.', '.', '7', '.', '.', '.', '.'],
  ['6', '.', '.', '1', '9', '5', '.', '.', '.'],
  ['.', '9', '8', '.', '.', '.', '.', '6', '.'],
  ['8', '.', '.', '.', '6', '.', '.', '.', '3'],
  ['4', '.', '.', '8', '.', '3', '.', '.', '1'],
  ['7', '.', '.', '.', '2', '.', '.', '.', '6'],
  ['.', '6', '.', '.', '.', '.', '2', '8', '.'],
  ['.', '.', '.', '4', '1', '9', '.', '.', '5'],
  ['.', '.', '.', '.', '8', '.', '.', '7', '9']
]
solveSudoku(board);
console.info(board);
